### ISSUE TITLE
Way to use Lazy?

### ISSUE THREAD

**Issue by CharlonTank** [nunntom/elm-ui-select#3]
Hello !

I've been using your select in production and it's working very nicely.

Only exception is when the select contains lots of items. The whole page becomes very slow. Any idea/best practice to fix that?

I will just add it to the readme of nunntom/elm-ui-select ;)

**Comment by nunntom** [comment]
Hello @CharlonTank 

Thanks for your feedback. I'm happy to hear you're using it and it's working well.

Can I ask how many items roughly do you have in the select when you notice this slowdown? Is the page slow all the time or just when interacting with the select? Any example code?

**Comment by CharlonTank** [comment]
Hey @nunntom  thank you for you fast answer!

22000 items 😂. Yeah it's slow all the time, not only when interacting with the select in question.
I use to `setItems` in the view, but I now `setItems` after fetching the items, but I feel like it doesn't change anything in terms of lag.

Here is the code where `model.aggregationSearchSelects.fullTextSelect.select` can be a list of 22000 items x)
```
, Select.withFilter (Just Select.startsWithThenContains)
    (Select.view [ height <| px 18, Border.rounded 0, moveUp 10 ]
        { onChange = SelectAggregatMsg "full-text-select"
        , label = Input.labelAbove [ width <| px 200, moveDown 2, moveLeft 23 ] <| text "Référence/Label"
        , placeholder = Nothing
        , itemToString = aggregationToString
        }
    )
    |> Select.withMenuAlwaysBelow
    |> Select.toElement model.aggregationSearchSelects.fullTextSelect.select
```

Is there a way to make the select filtering/showing results after that 3 characters has been entered?

**Comment by nunntom** [comment]
Hi @CharlonTank .

Yeah, 22000 items is way too many! I did play around with lazy and keyed but i didn't see a lot of difference.

When I have a large number of possible items I tend to be fetching them from an api using `Select.request` and ideally the back-end will filter, sort and limit the number of returned items based on the search query so I'm never dealing with too many on the front-end.

I could add a configuration option to the select to not show any items until a minimum number of chars has been entered. That might alleviate the problem but I expect there still might be a little slowdown just filtering that number of items. I'll look into doing this.

Btw `startsWithThenContains` is the default filter so you should be able to remove `Select.withFilter (Just Select.startsWithThenContains)`.

**Comment by CharlonTank** [comment]
Hey Nuuntom :) Thank you for your reply.

> Yeah, 22000 items is way too many! I did play around with lazy and keyed but i didn't see a lot of difference.

It's weird that lazy is not able to reduce this problem in front though.

> When I have a large number of possible items I tend to be fetching them from an api using Select.request and ideally the back-end will filter, sort and limit the number of returned items based on the search query so I'm never dealing with too many on the front-end.

Hey, yeah I thought of the same but the backend is in Scala and I'm not used to it and not really in charge. I'll need to understand and to create a route for that api call for sure.

> I could add a configuration option to the select to not show any items until a minimum number of chars has been entered. That might alleviate the problem but I expect there still might be a little slowdown just filtering that number of items. I'll look into doing this.

Thank you 

> Btw startsWithThenContains is the default filter so you should be able to remove Select.withFilter (Just Select.startsWithThenContains)

Ah yeah, thank you :)

As soon as I have updates on this, I let you know.

Have a good week!

**Comment by nunntom** [comment]
See v3.1.0 for [Select.withMinInputLength](https://package.elm-lang.org/packages/nunntom/elm-ui-select/3.1.0/Select#withMinInputLength)

Please let me know if this helps with your issue. @CharlonTank

**Comment by CharlonTank** [comment]
> See v3.1.0 for [Select.withMinInputLength](https://package.elm-lang.org/packages/nunntom/elm-ui-select/3.1.0/Select#withMinInputLength)
> 
> Please let me know if this helps with your issue. @CharlonTank

That's amazing thank you! Yeah it solved my problem!

**Comment by CharlonTank** [comment]
While `withMinInputLength` resolves my current issue, I think there is still room for optimization in the select component when not using that prop. For example, maybe implementing lazy loading for the items by only loading the items that are currently visible in the dropdown, rather than loading all the items at once. 
I'm unsure why the input text is causing lag, as I am only calling setItems when loading the large list and not in the view. @nunntom, are you available for a screenshare call to discuss further and explore these optimization opportunities?

**Comment by nunntom** [comment]
> While withMinInputLength resolves my current issue, I think there is still room for optimization in the select component when not using that prop. For example, maybe implementing lazy loading for the items by only loading the items that are currently visible in the dropdown, rather than loading all the items at once.

Hmm. It couldn't only "load" (add to vdom) the items visible in the menu, but potentially it could incrementally add more items as the scroll reaches the bottom. That wouldn't be too hard to implement.

> I'm unsure why the input text is causing lag, as I am only calling setItems when loading the large list and not in the view.

Is the input text causing lag for you now even when using `withMinInputLength` or are you talking about without it? I wouldn't be surprised if there is some lag because you're still running the filtering logic on that whole huge list on each keypress beyond the minimum, even though there's now less slowdown from the vdom diffing.

@CharlonTank

**Comment by nunntom** [comment]
Btw, I realised that without adding `withMinInputLength` there is a way you could have got similar behaviour. A bit hacky but assuming you aren't already using `Select.request`:

```elm
Select.updateWith
  [ Select.request
      (\input ->
          Ok (myFilterFunction input model.items) -- your own filter function
              |> Select.gotRequestResponse input
              |> Task.succeed
              |> Task.perform SelectMsg
      )
  ]
  SelectMsg
  selectMsg
  model.select
```

This would also debounce the input so that it doesn't perform the filtering logic on every keypress, but only when the user pauses typing. You can configure delay and min input with `Select.requestMinInputLength` and `Select.requestDebounceDelay`.

You need to write your own filter function though, keep the list of items to filter separately in your model, and make sure to pass `|> Select.withFilter Nothing` in view so it doesn't filter twice.

**Comment by CharlonTank** [comment]
> > I'm unsure why the input text is causing lag, as I am only calling setItems when loading the large list and not in the view.
> 
> Is the input text causing lag for you now even when using `withMinInputLength` or are you talking about without it? I wouldn't be surprised if there is some lag because you're still running the filtering logic on that whole huge list on each keypress beyond the minimum, even though there's now less slowdown from the vdom diffing.
> 
> @CharlonTank

No no, without `withMinInputLength`. But I'm talking about another select than this one or any other action the user would do. Why would the filtering happens at each update in the `update` function and not only when the user is currently filtering it?

**Comment by nunntom** [comment]
Yes, the filtering happens in the select view code, which means every time elm re-runs the view it will re-run the filtering on that huge list of yours which may happen when interacting with any other element on the page. It would be more efficient to filter in update but it would lose the ability to set items in the view which has been useful for some of my use-cases to avoid multiple sources of truth in the model. I will think about other ways round this.

In the meantime to mitigate this you can export the select view into a separate function and use `Element.Lazy` in your view. Then it will only run the filter when the select is updated directly. Unfortunately I can't make use of lazy internally because the view config object contains functions and so doesn't work with lazy. 

The [workaround above](#issuecomment-1400439580) would also work to move the filtering logic into update.

**Comment by nunntom** [comment]
@CharlonTank  Please try [v3.1.1](https://package.elm-lang.org/packages/nunntom/elm-ui-select/3.1.1/):

## 3.1.1

### Enhancements

- Optimised filtering. Previously filtering ran every time the virtual DOM re-rendered the select. This caused performance issues with large lists of items. Now the filter only runs when the input or options are changed.
- Filter startsWithThenContains now uses a faster implementation.

**Comment by CharlonTank** [comment]
> @CharlonTank  Please try [v3.1.1](https://package.elm-lang.org/packages/nunntom/elm-ui-select/3.1.1/):
> 
> 
> 
> ## 3.1.1
> 
> 
> 
> ### Enhancements
> 
> 
> 
> - Optimised filtering. Previously filtering ran every time the virtual DOM re-rendered the select. This caused performance issues with large lists of items. Now the filter only runs when the input or options are changed.
> 
> - Filter startsWithThenContains now uses a faster implementation.

I'll try that and let you know! Nice improvement!
